### ISSUE TITLE
How to create a new channel via API?

### ISSUE THREAD

**Issue by noho501** [datarhei/restreamer#409]
Hi,

I would like to create a new channel, video input ( RTMP Server ) then will import some PUBLICATION to the new channel through API without UI. Any ideas?

Thanks and best regards

<img width="1121" alt="create-channel" src="https://user-images.githubusercontent.com/35057759/188102147-59ff3693-0772-473d-b0b6-613ca5b0d05a.png">

<img width="549" alt="video-setup" src="https://user-images.githubusercontent.com/35057759/188102213-88af1455-995b-4c0f-9df2-f54209c75919.png">

<img width="822" alt="add-platforms" src="https://user-images.githubusercontent.com/35057759/188102653-ed7dd727-4458-4aed-b237-b8055e55fcd5.png">

**Comment by jstabenow** [comment]
Hey @noho501 
Should the processes be displayed in the user interface?

**Comment by noho501** [comment]
Hi @jstabenow 

It should be displayed in the UI, so the admin can manage them. Could you please help me with how to do that?

**Comment by jstabenow** [comment]
Yes, no problem. But it takes some time. So I will answer you in the coming days.

**Comment by noho501** [comment]
Thanks a lot

**Comment by noho501** [comment]
Hi @jstabenow Any updates?

**Comment by jstabenow** [comment]
> Hi @jstabenow Any updates?

I am working on a CLI with which you can read out, edit or create all processes, for example. 

That should help you with your project. And if you want to develop custom management.

There are API clients for [Go](https://github.com/datarhei/core-client-go) and [Python](https://github.com/datarhei/core-client-python). And for Javascript, [here](https://github.com/datarhei/restreamer-ui/blob/main/src/utils/api.js) is the file from the Restreamer UI we use for API communication.

As soon as the CLI is ready, I will let you know.

**Comment by noho501** [comment]
Hi, @jstabenow Thank you very much for your help.

**Comment by privatesam** [comment]
I'd be interested in the CLI when its ready too.  Thanks.

**Comment by seccentral** [comment]
Nice that this is already raised. 

I'm also interested in this, especially in creating a new channel from the cli or api for RTMP. 
Previously using nginx with rtmp module, all I had to do was define a location path ie `/live` and then publish my stream from cli like 
`ffmpeg [...] rtmp://<host>/<path>/{my_tag}" ` -> would be automatically ingested by nginx at /path/my_tag and an entry would be added to it's stat page.
Would love to do that here also, but at the moment restreamer needs manual gui setup for every new channel after which restreamer issues a `rtmp://<host>/<path>/<random-generated-string>.stream?token={my_token}` 

Funny thing is, ffmpeg can stream to restreamer using `rtmp://<host>/<path>/teststream?token={my_token}` but this stream is nowhere visible in the admin page, nor the publication page

**Comment by ioppermann** [comment]
@seccentral

> Funny thing is, ffmpeg can stream to restreamer using rtmp://.../teststream?token={my_token} but this stream is nowhere visible in the admin page, nor the publication page

This is true. You can choose any name for the stream and the RTMP server will accept it (as long as the token is correct). If you want to use it, you can enter the internal address for this stream as a network source. The internal address is the same address you use for pushing the stream, but replace the host with `127.0.0.1:1935` (or whatever RTMP port you configured).

The process in the GUI with suggesting a RTMP address with a random generated string and probing is for convenience, so you don't have to bother with the internal address of the RTMP server.

We'll take it as a suggestion to list available RTMP streams, regardless if the name has been suggested or you chose the name.

**Comment by seccentral** [comment]
> We'll take it as a suggestion to list available RTMP streams, regardless if the name has been suggested or you chose the name.

Would be super useful to see incoming streams in the admin and especially playersite with their names taken from the ffmpeg cli command, helps a lot with debugging and overviewing multiple streams in dev environments like computer vision and ai tinkering: push a stream from a devkit board and see it pop in the playersite automatically

**Comment by kyjus25** [comment]
Agreed. I am looking to reproduce the functionality of [Bambuser](https://bambuser.com/) with Restreamer. I have a Node script that converts a webcam WEBM stream from my website into an RTMP feed that Restreamer can ingest. I was surprised at how much configuration is involved in creating a new channel on demand through Restreamer's [API](https://demo.datarhei.com/api/swagger/index.html), and would very much appreciate an easy-to use `createChannel` endpoint that abstracts out all the JSON config bits associated with it. I got it to work by inspecting the UI's POST requests and mimicing the numerous endpoint calls, but a simple endpoint that takes in an aspect ratio and a stream URL would be much nicer.

**Comment by RLutsch** [comment]
sitting with exacly the same problem, be great if I could hit the api and say something like POST /api/channels then {data: channelID } etc etc? a cli or sdk could also do the trick

**Comment by pissten** [comment]
Is this implemented already? this thread is a couple of years old now, so just curious if it is possible to create a process/stream channel over api, that also is available through the UI?

**Comment by sandmonstersandy** [comment]
Hey @kyjus25 , would you be open to sharing the code/endpoint calls you made? I am about to start the same exercise.

**Comment by kyjus25** [comment]
@sandmonstersandy I did find the code, I still have it. I was able to reverse engineer the UI back then in order to make the same calls and it did end up creating a channel. It was for a now dead project and I'm embarrassed at the code stank to share it as-is.

I remember there were modals and polling going on in the UI and it would only work if I made the calls in the exact same order. A lot of it was hard-coding ffmpeg values that I didn't know anything about at the time.

I would be interested to revisit this. I'll see if the code still works, clean it up, and/or optimize and share it soon.

**Comment by kyjus25** [comment]
@sandmonstersandy Ok! Thanks for your patience, got it figured out.
Repo is located here for reference: https://github.com/kyjus25/restreamer-api. 
To start it:
1. Install [Bun](https://bun.sh/)
2. Create a `.env` file at the root with these values:
```bash
RESTREAMER_URL=http://localhost:8080
RESTREAMER_USERNAME=admin
RESTREAMER_PASSWORD=admin
RESTREAMER_RTMP_TOKEN=12345678 # created from the Restreamer UI when you enable RTMP
```
3. At the root of the project, run `bun install`
4. Run `bun dev` to start the Node backend
5. Swagger UI is served at `http://localhost:3000/swagger`

Hope it helps, best of luck in your project!

--- 

For creating an RTMP stream for OBS (assuming restreamer is also on localhost):
1. Generate a random UUIDv4, use this to build a stream link in this structure: `rtmp://localhost/<UUIDv4>.stream/<RTMP token created from UI>`. Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/restreamer.api.ts#L31-L41).
2. **Important for automation**: Start streaming to this URL before proceeding. It is needed in order to probe the stream for audio and video channels (the same as the UI does)
3. Log in to the API to get an access token (`[POST] /api/login`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/restreamer.api.ts#L14-L30).
4. Create a process (async ffmpeg task) that goes and probes the stream (`[POST] /api/v3/process`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/payload/probe_process.ts).
5. Check the response of the probe. It should return the video and audio tracks of your stream (`/api/v3/process/restreamer-ui%3Aingest%3A<UUIDv4>_probe/probe`). You'll need this info later. Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/restreamer.api.ts#L58-L73).
6. Delete the probe process once you have the tracks (`[DELETE] /api/v3/process/restreamer-ui%3Aingest%3A<UUIDv4>_probe`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/restreamer.api.ts#L74-L90).
7. Create the main channel process that is used to output an m3u8 file (`[POST] /api/v3/process`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/payload/channel_process.ts).
8. Update the UI with stream metadata like the tracks (probed earlier), stream name, description, author, and license (`[PUT] /api/v3/process/restreamer-ui%3Aingest%3A<UUIDv4>/metadata/restreamer-ui`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/payload/channel_metadata.ts).
9. Create a snapshot process. I think this is for the screenshots in the UI (`[POST] /api/v3/process`). Example [here](https://github.com/kyjus25/restreamer-api/blob/main/src/lib/payload/snapshot_process.ts).

https://github.com/user-attachments/assets/7fb3c141-6b4e-4d94-98df-d91133d54979
